frontend/apps/filemanager/filemanagerfilesearcher.lua

```diff
@@ -3,273 +3,333 @@ local CheckButton = require("ui/widget/checkbutton")
-    local sorting
-    if self.ui.file_chooser then
-        sorting = self.ui.file_chooser:getSortingFunction(collate, reverse_collate)
-    else
-        sorting = FileChooser:getSortingFunction(collate, reverse_collate)
-    end
-
-    table.sort(self.results, sorting)
-    self.search_menu:switchItemTable(T(_("Search results (%1)"), #self.results), self.results)
+    self.search_menu:switchItemTable(T(_("Search results (%1)"), #results), results)
     UIManager:show(menu_container)
+    if self.no_metadata_count ~= 0 then
+        self:showSearchResultsMessage()
+    end
 end
 
-function FileSearcher:onMenuHold(item)
-    local ReaderUI = require("apps/reader/readerui")
-    local is_file = item.attr.mode == "file"
-    local fullpath = item.dir .. "/" .. item.name .. (is_file and "" or "/")
-    local buttons = {
-        {
+function FileSearcher:onMenuSelect(item)
+    local dialog
+    local buttons = {}
+    if item.is_file then
+        table.insert(buttons, {
             {
-                text = _("Cancel"),
+                text = _("Book information"),
                 callback = function()
-                    UIManager:close(self.results_dialog)
+                    UIManager:close(dialog)
+                    FileManagerBookInfo:show(item.path)
                 end,
             },
             {
-                text = _("Show folder"),
+                text = _("Open"),
+                enabled = DocumentRegistry:hasProvider(item.path),
                 callback = function()
-                    UIManager:close(self.results_dialog)
+                    UIManager:close(dialog)
                     self.close_callback()
-                    self._manager:showFolder(fullpath)
+                    require("apps/reader/readerui"):showReader(item.path)
                 end,
             },
+        })
+    end
+    table.insert(buttons, {
+        {
+            text = _("Cancel"),
+            callback = function()
+                UIManager:close(dialog)
+            end,
         },
-    }
-    if is_file then
-        table.insert(buttons[1], {
-            text = _("Open"),
+        {
+            text = _("Show folder"),
             callback = function()
-                UIManager:close(self.results_dialog)
+                UIManager:close(dialog)
                 self.close_callback()
-                ReaderUI:showReader(fullpath)
+                self._manager:showFolder(item.path)
             end,
-        })
-    end
-
-    self.results_dialog = ButtonDialogTitle:new{
-        title = fullpath,
+        },
+    })
+    dialog = ButtonDialogTitle:new{
+        title = item.path,
         buttons = buttons,
     }
-    UIManager:show(self.results_dialog)
+    UIManager:show(dialog)
+end
+
+function FileSearcher:onMenuHold(item)
+    if item.is_file then
+        if DocumentRegistry:hasProvider(item.path) then
+            UIManager:scheduleIn(0.5, function()
```

Comment:
Worth a comment - so we remember that and may be one day find a better solution to it.
`-- Workaround an issue with this long-press' release being handled as tap by the Reader we launch`

Reply:
Actually not a good workaround, if a long-press is longer than 0.5 sec it doesn't help.

Reply:
Not good, but good enough ? :) Let's not block this PR, we can fix that later if we get a better idea.

Another idea (dunno if it is good).
```diff
--- a/frontend/ui/widget/menu.lua
+++ b/frontend/ui/widget/menu.lua
@@ -131,7 +131,7 @@ function MenuItem:init()
         },
         HoldSelect = {
             GestureRange:new{
-                ges = "hold",
+                ges = self.handle_hold_on_hold_release and "hold_release" or "hold",
                 range = self.dimen,
             },
         },
```
and provide and pass `handle_hold_on_hold_release` when you know it's best to handle it that way.

I think we mostly everywhere handle long-press before the finger lift - but we have a few places (don't really remember where) where we handle it on finger lift (so, hold_release), probably for good reasons. I remember I had sometimes been surprised by that (waiting, and seeing nothing done, until I lift my finger), and it's a bit inconsistent, but well, it's just fine.